File: Data Structures & Algorithms/target-sum/submission-2.cpp
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) {
        map<pair<int,int>, int> dp; // (index, cur_sum) - > number of ways

        return backtrack(0, 0, nums, dp, target);
    }

    int backtrack(int i, int cur_sum, vector<int>& nums, map<pair<int,int>, int>& dp, int target){
        if(dp.count({i,cur_sum})){
            return dp[{i, cur_sum}];
        }

        if(i==nums.size()){
            return cur_sum == target ? 1 : 0;
        }

        dp[{i, cur_sum}] = ( backtrack(i+1, cur_sum + nums[i], nums, dp, target) + backtrack(i+1, cur_sum - nums[i], nums, dp, target) );

        return dp[{i, cur_sum}];
    }
};

File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        ns = len(nums)
        cache = [None]*(ns)

        def dp(n):
            if n>=ns:
                return 0
            if cache[n] is not None:
                return cache[n]
            
            cache[n] = max(nums[n]+dp(n+2),dp(n+1))

            return cache[n]
        
        return dp(0)